### PR TITLE
Adds cache to allow downloads do pause and restart

### PR DESCRIPTION
**Pull request by cuducos**
The suggested cache works as follows:

`chunk`, once integrated with this cache system, will create a directory `.cache` and some files inside it:
* `.cache/.chunk_size` with just the size of the chunk, e.g. `8192`;
* a file per file being downloaded, e.g. if we're downloading `a.zip` and `b.zip` `.cache/a.zip` and `.cache/b.zip`.

These last files will contain a sequence of `0` and `1`, meaning which chunks are downloaded (`1`) or not (`0`). For example, if a file needs four chunks, and the second one is done, the cache file's contents will be `0100`.

Every time a new cache is created, it checks for these files, and if the chunk size matches, it loads the info (which chunks are done/pending) from the cache.

### CODE REVIEW

**Comment by cuducos** [comment]
Just as a reminder, once the main branch is downloading chunk by chunk (see #5), we can integrate that functionality (see #8).

**Review by danielfireman — CHANGES_REQUESTED** [review]
Few things:

1. as this is a download manager, I see this more like a progress tracker than actually a cache. But this is a super personal view, please feel free to ignore this comment.
2. Avoid having a single file to store information about all downloads in progress. The main reason is to avoid the case of multiple chunk instances modifying the progress file, which might cause hard-to-spot problems.

**Comment by danielfireman** [comment]
Finally, as a reviewer, it is hard to grasp and contribute to such a complex piece of substrate infrastructure without knowing the use cases. My suggestion for next time: we could have the usage and the substrate implementation in the same PR. That means we would not send a PR with all the substrate infrastructure (complete functionality) but instead split it into multiple PRs along with their usage.

**Comment by cuducos** [comment]
A lot to comment, but the main point I don't get are this:

> Avoid having a single file to store information about all downloads in progress

And this:

> As suggested, each file should have one cache file.

I'll focus on one big issue at a time, so this doesn't mean I'm ignoring all the other valuable comments. And I'll prevent writing more code before going all your (again) valuable comments : )

---

My point to start is: I think you got it wrong on this part of the files architecture. As I explained in the PR description:

> `chunk`, once integrated with this cache system, will create a directory `.cache` and some files inside it:
> 
>   *  `.cache/.chunk_size` with just the size of the chunk, e.g. `8192`;
>   *  a file per file being downloaded, e.g. if we're downloading `a.zip` and `b.zip`, `.cache/a.zip` and `.cache/b.zip`.

This means that each file being downloaded has it's own file:
* `.cache/a.zip` knows which chunks of `a.zip` are already downloaded (and knows nothing about chunk size, or about `b.zip` download)
* `.cache/b.zip` knows which chunks of `b.zip` are already downloaded (and knows nothing about chunk size, or about `a.zip` download)
 
 **Example**
 
 _At start_
 
 ```console
 $ chunk http://whatever.com/a.zip http://whatever.com/b.zip
 ```
 
 Let's say each of these files will need, respectively, 8 and 3 chunks to complete.

 This will create three files:
 * `.cache/.chunk_size` with the contents `8192`
 * `.cache/a.zip` with the contents `00000000`
 * `.cache/b.zip` with the contents `000`

_After a minute_

Let's say after a minute, chunks 1, 3 and 6 are done for `a.zip` and 2 and 3 are done for `b.zip`. Our files would look like this:
 * `.cache/.chunk_size` will be unchanged since de chunk size is constant during the download
 * `.cache/a.zip` with the contents `10100100`
 * `.cache/b.zip` with the contents `011`
 
 Does that make sense and is aligned with your view for this feature or is it me who got it wrong (which often happens, surely hahaha…)?

**Comment by danielfireman** [comment]
> Does that make sense and is aligned with your view for this feature or is it me who got it wrong (which often happens, surely hahaha…)?

Super! My only addition is that the `.chunk_size` file is not necessary. It is actually dangerous, as it could be written and read by many processes.

Just noticed another important thing: the file path also must be persisted, as the user can adjust it between downloads.

Thanks for such an awesome example! I am going to use it for summarising my main suggestions.

**Example**

_At start_

`$ chunk http://whatever.com/a.zip http://whatever.com/b.zip`

Let's say each of these files will need, respectively, 8 and 3 chunks to complete.

This will create two files:

.chunk/a.zip with the contents: path_to_the_actual_file_content::8192::00000000
.chunk/b.zip with the contents: path_to_the_actual_file_content::8192::000

The `::` can be replaced by any separator you prefer. Or it can be a binary version of a data structure (`gob`). 

_After a minute_

Let's say after a minute, chunks 1, 3 and 6 are done for a.zip and 2 and 3 are done for b.zip. Our files would look like this:

.chunk/a.zip with the contents path_to_the_actual_file_content::8192::10100100
.chunk/b.zip with the contents path_to_the_actual_file_content::8192::011

With that, each file being downloaded owns its own configuration file, which means each download can safely be manipulated by different chunk processes.

**Comment by cuducos** [comment]
Got it! Gonna refactor it this way, renaming stuff using something like `progress` or `tracking`, adding more comments and then we can check what is still remaining for the previous topics.

My other doubts seem irrelevant now because I fell that with the refactor those parts would be gone. Let's see : ) Thank you so much!

File: progress.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	if !ok {
+		return nil
+	}
+	if err := os.Remove(p.path); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("error cleaning up progress files %s: %w", p.path, err)
+	}
+	return nil // Either not empty or error, suits both cases
+}
+
+func newProgress(path, url string, chunkSize uint64, chunks int, restart bool) (*progress, error) {
```
Comment (danielfireman) [inline progress.go:126]: [API suggestion] have two different methods:

- newProgress(path, url string, chunkSize uint64, chunks int) (*progress, error)
- restartProgress(path, url string, chunkSize uint64, chunks int) (*progress, error)

The only difference here is the absence of the boolean restart. I feel this parameter makes the API very error-prone. It is totally personal; thus, feel free to ignore it.

**Review by danielfireman — APPROVED** [review]
One small comment follows below.

File: progress.go (hunk 1 above)
Comment (cuducos) [reply progress.go:126]: Got the point. We can come back and refactor when we integrate that with the downloader : ) I have the impression the downloader should be agnostic about the existence or not of an on going _progress_… but let's keep it in mind.

**Pull request merged**